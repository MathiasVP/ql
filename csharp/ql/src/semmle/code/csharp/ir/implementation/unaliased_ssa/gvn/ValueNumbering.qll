private import internal.ValueNumberingInternal
private import internal.ValueNumberingImports

/**
 * Provides additional information about value numbering in IR dumps.
 */
class ValueNumberPropertyProvider extends IRPropertyProvider {
  override string getInstructionProperty(Instruction instr, string key) {
    exists(ValueNumber vn |
      vn = valueNumber(instr) and
      key = "valnum" and
      if strictcount(vn.getAnInstruction()) > 1 then result = vn.toString() else result = "unique"
    )
  }
}

/**
 * The value number assigned to a particular set of instructions that produce equivalent results.
 */
class ValueNumber extends TValueNumber {
  final string toString() { result = "GVN" }

  final string getDebugString() {
    result = "ValueNumber: " +
        strictconcat(this.getAnInstruction().getUnconvertedResultExpression().toString(), ", ")
  }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    this
        .getAnInstruction()
        .getLocation()
        .hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  /**
   * Gets the instructions that have been assigned this value number. This will always produce at
   * least one result.
   */
  final Instruction getAnInstruction() { this = valueNumber(result) }

  /**
   * Gets one of the instructions that was assigned this value number. The chosen instuction is
   * deterministic but arbitrary. Intended for use only in debugging.
   */
  final Instruction getExampleInstruction() {
    result = min(Instruction instr |
        instr = getAnInstruction()
      |
        instr order by instr.getBlock().getDisplayIndex(), instr.getDisplayIndexInBlock()
      )
  }

  /**
   * Gets an `Operand` whose definition is exact and has this value number.
   */
  final Operand getAUse() { this = valueNumber(result.getDef()) }

  final string getKind() {
    this instanceof TVariableAddressValueNumber and result = "VariableAddress"
    or
    this instanceof TInitializeParameterValueNumber and result = "InitializeParameter"
    or
    this instanceof TInitializeThisValueNumber and result = "InitializeThis"
    or
    this instanceof TStringConstantValueNumber and result = "StringConstant"
    or
    this instanceof TFieldAddressValueNumber and result = "FieldAddress"
    or
    this instanceof TBinaryValueNumber and result = "Binary"
    or
    this instanceof TPointerArithmeticValueNumber and result = "PointerArithmetic"
    or
    this instanceof TUnaryValueNumber and result = "Unary"
    or
    this instanceof TInheritanceConversionValueNumber and result = "InheritanceConversionr"
    or
    this instanceof TUniqueValueNumber and result = "Unique"
  }

  Language::Expr getAnExpr() { result = getAnInstruction().getUnconvertedResultExpression() }
}

/**
 * Gets the value number assigned to `instr`, if any. Returns at most one result.
 */
ValueNumber valueNumber(Instruction instr) { result = tvalueNumber(instr) }

/**
 * Gets the value number assigned to the exact definition of `op`, if any.
 * Returns at most one result.
 */
ValueNumber valueNumberOfOperand(Operand op) { result = valueNumber(op.getDef()) }
