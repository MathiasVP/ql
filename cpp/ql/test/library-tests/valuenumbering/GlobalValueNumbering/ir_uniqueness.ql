import cpp
import semmle.code.cpp.ir.ValueNumbering
import semmle.code.cpp.ir.IR

// Unlike the AST global value numbering, which produces *exactly* 1 value number per expression,
// the IR global value numbering produces *at most* 1 value number per instruction.
from Instruction i
where count(valueNumber(i)) != 1
select i, concat(ValueNumber g | g = valueNumber(i) | g.getKind(), ", ")
