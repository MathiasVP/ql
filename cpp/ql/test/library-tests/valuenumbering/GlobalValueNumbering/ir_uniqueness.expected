| test.cpp:1:5:1:10 | AliasedUse: test00 |  |
| test.cpp:1:5:1:10 | EnterFunction: test00 |  |
| test.cpp:1:5:1:10 | ExitFunction: test00 |  |
| test.cpp:1:5:1:10 | ReturnValue: test00 |  |
| test.cpp:1:5:1:10 | UnmodeledUse: test00 |  |
| test.cpp:8:1:8:1 | NoOp: return ... |  |
| test.cpp:12:5:12:10 | AliasedUse: test01 |  |
| test.cpp:12:5:12:10 | EnterFunction: test01 |  |
| test.cpp:12:5:12:10 | ExitFunction: test01 |  |
| test.cpp:12:5:12:10 | ReturnValue: test01 |  |
| test.cpp:12:5:12:10 | UnmodeledUse: test01 |  |
| test.cpp:19:1:19:1 | NoOp: return ... |  |
| test.cpp:25:5:25:10 | AliasedUse: test02 |  |
| test.cpp:25:5:25:10 | EnterFunction: test02 |  |
| test.cpp:25:5:25:10 | ExitFunction: test02 |  |
| test.cpp:25:5:25:10 | ReturnValue: test02 |  |
| test.cpp:25:5:25:10 | UnmodeledUse: test02 |  |
| test.cpp:30:3:30:17 | Call: call to change_global02 |  |
| test.cpp:33:1:33:1 | NoOp: return ... |  |
| test.cpp:39:5:39:10 | AliasedUse: test03 |  |
| test.cpp:39:5:39:10 | EnterFunction: test03 |  |
| test.cpp:39:5:39:10 | ExitFunction: test03 |  |
| test.cpp:39:5:39:10 | ReturnValue: test03 |  |
| test.cpp:39:5:39:10 | UnmodeledUse: test03 |  |
| test.cpp:47:1:47:1 | NoOp: return ... |  |
| test.cpp:49:14:49:22 | AliasedUse: my_strspn |  |
| test.cpp:49:14:49:22 | EnterFunction: my_strspn |  |
| test.cpp:49:14:49:22 | ExitFunction: my_strspn |  |
| test.cpp:49:14:49:22 | ReturnValue: my_strspn |  |
| test.cpp:49:14:49:22 | UnmodeledUse: my_strspn |  |
| test.cpp:53:10:53:21 | ConditionalBranch: ... != ... |  |
| test.cpp:56:13:56:24 | ConditionalBranch: ... != ... |  |
| test.cpp:56:31:56:42 | ConditionalBranch: ... != ... |  |
| test.cpp:59:9:59:20 | ConditionalBranch: ... == ... |  |
| test.cpp:60:7:60:12 | NoOp: break; |  |
| test.cpp:63:3:63:3 | NoOp: label ...: |  |
| test.cpp:75:6:75:11 | AliasedUse: test04 |  |
| test.cpp:75:6:75:11 | EnterFunction: test04 |  |
| test.cpp:75:6:75:11 | ExitFunction: test04 |  |
| test.cpp:75:6:75:11 | ReturnVoid: test04 |  |
| test.cpp:75:6:75:11 | UnmodeledUse: test04 |  |
| test.cpp:79:7:79:33 | ConditionalBranch: ... < ... |  |
| test.cpp:79:17:79:20 | FieldAddress: val1 | FieldAddress, Unary |
| test.cpp:79:30:79:33 | FieldAddress: val2 | FieldAddress, Unary |
| test.cpp:82:1:82:1 | NoOp: return ... |  |
| test.cpp:84:6:84:11 | AliasedUse: test05 |  |
| test.cpp:84:6:84:11 | EnterFunction: test05 |  |
| test.cpp:84:6:84:11 | ExitFunction: test05 |  |
| test.cpp:84:6:84:11 | ReturnVoid: test05 |  |
| test.cpp:84:6:84:11 | UnmodeledUse: test05 |  |
| test.cpp:88:7:88:12 | ConditionalBranch: ... != ... |  |
| test.cpp:89:1:89:1 | NoOp: return ... |  |
| test.cpp:91:5:91:21 | AliasedUse: regression_test00 |  |
| test.cpp:91:5:91:21 | EnterFunction: regression_test00 |  |
| test.cpp:91:5:91:21 | ExitFunction: regression_test00 |  |
| test.cpp:91:5:91:21 | ReturnValue: regression_test00 |  |
| test.cpp:91:5:91:21 | UnmodeledUse: regression_test00 |  |
| test.cpp:104:5:104:26 | AliasedUse: inheritanceConversions |  |
| test.cpp:104:5:104:26 | EnterFunction: inheritanceConversions |  |
| test.cpp:104:5:104:26 | ExitFunction: inheritanceConversions |  |
| test.cpp:104:5:104:26 | ReturnValue: inheritanceConversions |  |
| test.cpp:104:5:104:26 | UnmodeledUse: inheritanceConversions |  |
| test.cpp:105:15:105:15 | FieldAddress: b | FieldAddress, Unary |
| test.cpp:107:15:107:15 | FieldAddress: b | FieldAddress, Unary |
| test.cpp:112:6:112:11 | AliasedUse: test06 |  |
| test.cpp:112:6:112:11 | EnterFunction: test06 |  |
| test.cpp:112:6:112:11 | ExitFunction: test06 |  |
| test.cpp:112:6:112:11 | ReturnVoid: test06 |  |
| test.cpp:112:6:112:11 | UnmodeledUse: test06 |  |
| test.cpp:117:1:117:1 | NoOp: return ... |  |
